src/dmd/backend/elpicpie.d

```diff
@@ -478,531 +826,182 @@ private elem *el_picvar(Symbol *s)
+                 *   mov EAX,FS:[RAX]
+                 */
+                Obj.refGOTsym();
+                tym_t tym = e.Ety;
+                e.Ety = TYsptr;         // TYsptr means FS:
 
-                e.Eoper = OPind;
-                e.EV.E1 = el_bint(OPadd,e1.ET,e1,e2);
-                e.EV.E2 = null;
+                e = el_una(OPind, tym, e);
+                break;
             }
+            default:
+                break;
         }
+    }
+    else
+    {
+        switch (s.Sclass)
+        {
+            case SCstatic:
+            case SClocstat:
+            case SCglobal:
                 break;
 
-            case mTYthread | mTYimport:
-                version (SCPP_HTOD) { } else assert(0);
-                tx86err(EM_thread_and_dllimport,s.Sident.ptr);     // can't be both thread and import
-                break;
+            case SCcomdat:
+            case SCcomdef:
+            case SCextern:
+            {
+                /* Generate:
+                 *   mov EAX,extern_tls@TLS_GOTIE[ECX]
+                 *   mov EAX,GS:[EAX]
+                 */
+                tym_t tym = e.Ety;
+                e.Eoper = OPrelconst;
+                e.Ety = TYnptr;
 
+                e = el_bin(OPadd, TYnptr, e, el_var(el_alloc_localgot()));
+                e = el_una(OPind, TYsptr, e);
+                e = el_una(OPind, tym, e);
+                break;
+            }
             default:
                 break;
         }
-}
     }
-    else
-        e.Ety = s.ty();
     return e;
 }
-}
 
-/**************************
- * Make a pointer to a `Symbol`.
- * Params: s = symbol
- * Returns: `elem` with address of `s`
+/**********************************************
+ * Create an address for TLS variable `s`.
+ * Use PIE protocol.
+ * Params: s = variable's symbol
+ * Returns: elem created
  */
-
-elem * el_ptr(Symbol *s)
+private elem *el_pieptr(Symbol *s)
 {
-    elem *e;
+    int x;
 
-    //printf("el_ptr(s = '%s')\n", s.Sident.ptr);
-    //printf("el_ptr\n");
+    //printf("el_pieptr(s = '%s')\n", s.Sident.ptr);
     symbol_debug(s);
     type_debug(s.Stype);
+    auto e = el_calloc();
+    e.Eoper = OPrelconst;
+    e.EV.Vsym = s;
+    e.Ety = TYnptr;
 
-    static if (TARGET_OSX)
+    elem* e0 = el_una(OPind, TYsize, el_long(TYsptr, 0)); // I64: FS:[0000], I32: GS:[0000]
+
+    if (I64)
     {
-        if (config.flags3 & CFG3pic && tyfunc(s.ty()) && I32)
+        Obj.refGOTsym();    // even though not used, generate reference to _GLOBAL_OFFSET_TABLE_
+        switch (s.Sclass)
         {
-            /* Cannot access address of code from code.
-             * Instead, create a data variable, put the address of the
-             * code in that data variable, and return the elem for
-             * that data variable.
-             */
-            Symbol *sd = symboldata(Offset(DATA), TYnptr);
-            sd.Sseg = DATA;
-            Obj.data_start(sd, _tysize[TYnptr], DATA);
```

Comment:
This line was removed. Is this deliberate?

Reply:
No. Good catch!

Reply:
And even if it was deliberate, I'd never do that for a refactoring advertised as a cut&paste job. It'd be a separate PR.